### PR TITLE
AST to Expression logic

### PR DESCRIPTION
**Pull request by cjllanwarne**
Logic to convert draft-3 AstNodes into wdlom expression elements

### CODE REVIEW

File: wdl/transforms/draft3/src/main/scala/wdl/draft3/transforms/ast2wdlom/AstNodeToExpressionElement.scala [hunk 1]
```diff
@@ -14,26 +17,196 @@ import wom.values._
...
   def convert(astNode: AstNode): ErrorOr[ExpressionElement] = astNode match {
 
     case t: Terminal if asPrimitive.isDefinedAt((t.getTerminalStr, t.getSourceString)) => asPrimitive((t.getTerminalStr, t.getSourceString)).map(PrimitiveLiteralExpressionElement)
-    case t: Terminal => s"No rule available to create ExpressionElement from terminal: ${t.getTerminalStr} ${t.getSourceString}".invalidNel
+    case t: Terminal if t.getTerminalStr == "identifier" => IdentifierLookup(t.getSourceString).validNel
+
+    case a: Ast if lhsRhsOperators.contains(a.getName) => useValidatedLhsAndRhs(a, lhsRhsOperators(a.getName))
+    case a: Ast if unaryOperators.contains(a.getName) => a.getAttributeAs[ExpressionElement]("expression").map(unaryOperators(a.getName)).toValidated
+    case a: Ast if a.getName == "TupleLiteral" => (a.getAttributeAsVector[ExpressionElement]("values") flatMap {
+      case pair if pair.length == 2 => PairLiteral(pair.head, pair(1)).validNelCheck
```
Comment (jbuabbud) [inline wdl/transforms/draft3/src/main/scala/wdl/draft3/transforms/ast2wdlom/AstNodeToExpressionElement.scala]: Why not pair.tail instead of pair(1)?

Comment (cjllanwarne) [reply wdl/transforms/draft3/src/main/scala/wdl/draft3/transforms/ast2wdlom/AstNodeToExpressionElement.scala]: `pair.tail` will return a `Vector[ExpressionElement]` representing everything except the `head`. But I only want the second element of the original Vector.

I could do `pair.tail.head` but it's easier to just go straight to the item at index 1

Comment (jbuabbud) [reply wdl/transforms/draft3/src/main/scala/wdl/draft3/transforms/ast2wdlom/AstNodeToExpressionElement.scala]: Makes sense! Thanks for the explanation.

File: wdl/transforms/draft3/src/main/scala/wdl/draft3/transforms/ast2wdlom/AstNodeToExpressionElement.scala [hunk 2]
```diff
@@ -14,26 +17,196 @@ import wom.values._
...
+
+  private def validateTwoParamEngineFunction(elementMaker: (ExpressionElement, ExpressionElement) => ExpressionElement, functionName: String)
+                                            (params: Vector[ExpressionElement]): ErrorOr[ExpressionElement] =
+    if (params.size == 2) {
+      elementMaker.apply(params.head, params(1)).validNel
+    } else {
+      s"Function $functionName expects exactly 2 arguments but got ${params.size}".invalidNel
+    }
+
+  private def validateThreeParamEngineFunction(elementMaker: (ExpressionElement, ExpressionElement, ExpressionElement) => ExpressionElement, functionName: String)
```
Comment (jbuabbud) [inline wdl/transforms/draft3/src/main/scala/wdl/draft3/transforms/ast2wdlom/AstNodeToExpressionElement.scala]: Is three the maximum number of parameters taken by an engine function?

Comment (cjllanwarne) [reply wdl/transforms/draft3/src/main/scala/wdl/draft3/transforms/ast2wdlom/AstNodeToExpressionElement.scala]: I guess in theory they could take any number but right now, there are no engine functions which take more than 3

File: wdl/transforms/draft3/src/test/cases/input_expressions.nowom.wdl [hunk 3]
```diff
@@ -1,8 +1,48 @@
...
+# This WDL contains expressions that won't be linkable, but it should be parseable and wdlom-able:
 workflow input_expressions {
   input {
-    Int four = 2 + 2
-#    Int i = [[0]][0][0]
+    Int ten = 5 + 5
+    Int zero = 5 - 5
+    Int twentyfive = 5 * 5
+    Int one = 5 / 5
+    Int zeroagain = 5 % 5
```
Comment (jbuabbud) [inline wdl/transforms/draft3/src/test/cases/input_expressions.nowom.wdl]: 🤔

**Pull request merged**